File: ProjectEulerPython/problems/problem_040.py
from problems.problem import Problem

class DigitIndexInformation:

  def __init__(self, digit: int, number: int, number_index):
    self.digit = digit
    self.number = number
    self.number_index = number_index

def number_digit(n: int, d: int) -> int:
  return int(str(n)[d])

def digit_index_to_information(i: int) -> DigitIndexInformation:
  d = 0
  n = 0
  ni = 0
  if i < 10: # 1 to 10; digits: 1
    n = i
    ni = 0
    d = i
  elif i < 190: # 10 to 189; digits: 2
    n = i // 2 + 5
    ni = i % 2
    d = number_digit(n, ni)
  elif i < 2890: # 190 to 2889; digits: 3
    n = (i - 1) // 3 + 37
    ni = (i - 1) % 3
    d = number_digit(n, ni)
  elif i < 38890: # 2890 to 38889; digits: 4
    n = (i - 2) // 4 + 278
    ni = (i - 2) % 4
    d = number_digit(n, ni)
  elif i < 488890: # 38890 to 488889; digits: 5
    n = i // 5 + 2222
    ni = i % 5
    d = number_digit(n, ni)
  elif i < 5888890: # 488890 to 5888889; digits: 6
    n = (i - 4) // 6 + 18519
    ni = (i - 4) % 6
    d = number_digit(n, ni)
  elif i < 68888890: # 5888890 to 68888889; digits: 7
    n = i // 7 + 158730
    ni = i % 7
    d = number_digit(n, ni)
  else:
    raise ValueError(f"out of range: i = {i}")
  return DigitIndexInformation(d, n, ni)

class Problem040(Problem):

  def calculate_answer(self) -> int:
    digit_indices = [1, 10, 100, 1000, 10000, 100000, 1000000]

    answer = 1
    digits = []
    for i in digit_indices:
      info = digit_index_to_information(i)
      answer *= info.digit
      digits.append(info.digit)
      self.print_detail(f"d_{i} = {info.digit} (number = {info.number}; index = {info.number_index + 1})")
    self.print_detail(f"{answer} = {' * '.join([str(d) for d in digits])}")

    return answer

File: ProjectEuler/Problems/Problem040.cs
using System;
using System.Collections.Generic;

namespace ProjectEuler.Problems
{
    public class Problem040 : Problem<long>
    {
        protected override long CalculateAnswer()
        {
            var answer = 1L;
            var digits = new List<int>();
            foreach (var index in digitIndices)
            {
                var info = DigitIndexToInformation(index);
                answer *= info.Digit;
                digits.Add(info.Digit);
                WriteLineDetail($"d_{index} = {info.Digit} (number = {info.Number}; index = {info.NumberIndex + 1})");
            }
            WriteLineDetail($"{answer} = {string.Join(" * ", digits)}");
            return answer;
        }

        private readonly int[] digitIndices =
        {
            1,
            10,
            100,
            1000,
            10000,
            100000,
            1000000,
        };

        private static DigitIndexInformation DigitIndexToInformation(long index)
        {
            int digit;
            long number;
            int numberIndex;
            if (index < 10) // 1 to 10; digits: 1
            {
                number = index;
                numberIndex = 0;
                digit = (int)index;
            }
            else if (index < 190) // 10 to 189; digits: 2
            {
                number = index / 2 + 5;
                numberIndex = (int)(index % 2L);
                digit = NumberDigit(number, numberIndex);
            }
            else if (index < 2890) // 190 to 2889; digits: 3
            {
                number = (index - 1) / 3 + 37;
                numberIndex = (int)((index - 1) % 3L);
                digit = NumberDigit(number, numberIndex);
            }
            else if (index < 38890) // 2890 to 38889; digits: 4
            {
                number = (index - 2) / 4 + 278;
                numberIndex = (int)((index - 2) % 4L);
                digit = NumberDigit(number, numberIndex);
            }
            else if (index < 488890) // 38890 to 488889; digits: 5
            {
                number = index / 5 + 2222;
                numberIndex = (int)(index % 5L);
                digit = NumberDigit(number, numberIndex);
            }
            else if (index < 5888890) // 488890 to 5888889; digits: 6
            {
                number = (index - 4) / 6 + 18519;
                numberIndex = (int)((index - 4) % 6L);
                digit = NumberDigit(number, numberIndex);
            }
            else if (index < 68888890) // 5888890 to 68888889; digits: 7
            {
                number = index / 7 + 158730;
                numberIndex = (int)(index % 7L);
                digit = NumberDigit(number, numberIndex);
            }
            else
            {
                throw new ArgumentOutOfRangeException(nameof(index));
            }
            return new DigitIndexInformation(digit, number, numberIndex);
        }

        private static int NumberDigit(long number, int digit)
        {
            return number.ToString()[digit] - '0';
        }

        private class DigitIndexInformation
        {
            public int Digit { get; }
            public long Number { get; }
            public int NumberIndex { get; }

            public DigitIndexInformation(int digit, long number, int numberIndex)
            {
                Digit = digit;
                Number = number;
                NumberIndex = numberIndex;
            }
        }
    }
}
